src/qibolab/instruments/rfsoc.py

```diff
@@ -33,15 +33,20 @@ def convert_qubit(qubit: Qubit) -> rfsoc.Qubit:
     return rfsoc.Qubit(0.0, None)
 
 
-def replace_pulse_shape(rfsoc_pulse: rfsoc.Pulse, shape: PulseShape) -> rfsoc.Pulse:
-    """Set pulse shape parameters in rfsoc pulse object."""
-    new = copy.copy(rfsoc_pulse)
-    shape_name = new.shape = shape.name.lower()
+def replace_pulse_shape(rfsoc_pulse: rfsoc_pulses.Pulse, shape: PulseShape) -> rfsoc_pulses.Pulse:
+    """Set pulse shape parameters in rfsoc_pulses pulse object."""
+
+    pulse_dict = asdict(rfsoc_pulse)
+
+    shape_name = shape.name.lower()
+    cls = rfsoc_pulses.Shape[shape_name].value
+    args = {}
     if shape_name in {"gaussian", "drag"}:
-        new.rel_sigma = shape.rel_sigma
+        args["rel_sigma"] = shape.rel_sigma
         if shape_name == "drag":
-            new.beta = shape.beta
-    return new
+            args["beta"] = shape.beta
+
+    return cls(**pulse_dict, **args)
```

Comment:
I am not so convinced with this function.
I try to simplify it and this is what I came up with:

```py
def replace_pulse_shape(rfsoc_pulse: rfsoc_pulses.Pulse, shape: PulseShape) -> rfsoc_pulses.Pulse:
    """Set pulse shape parameters in rfsoc_pulses pulse object."""


    new_pulse = getattr(rfsoc_pulses, shape.name)(**asdict(rfsoc_pulse))
    new_pulse.shape = shape.name.lower()
    if shape.name in {"Gaussian", "Drag"}:
        new_pulse.rel_sigma = shape.rel_sigma
        if shape.name == "Drag":
            new_pulse.beta = shape.beta

    return new_pulse
```
which works until you keep the same naming convention between rfsoc.pulses and PulseShape in qibolab. With this I think that you should also be able to drop the `Enum` in qibosoq and eventually the `shape` attribute since the information about the shape is already included in the class name.
This is just a suggestion. Let me know what you think :)

Reply:
I like it! (and already implemented it).

I'm not sure of dropping the Enum: I like that all the shapes are in a common collections... But maybe in the future it will appear useless.

(I also added a test for replace_pulse_shape, that was missing)

Reply:
Sure, we can keep also the `Enum` even if it is not used by this function.
Btw, I think that here https://github.com/qiboteam/qibolab/pull/510/commits/3efa5c82ecf0c9f186e6cf7e70668a9e9ca2c950 you forgot to update the `pulse.shape` attribute.

Reply:
If you mean `new_pulse.shape` it is not needed because it's already the default in the respective shape classes

Reply:
I think it is my fault.
This function is meant to be used only to go from `rfsoc_pulses.Pulse` to whatever pulses, right?
I had in mind another example where you can change the shape from a `Gaussian` to a `Rectangular`.
In this case the current implementation does not work (not only because of  `shape` but also because we will need to remove / add certain attributes depending on the conversion).
If you are planning to use it only in the first case it should be fine (eventually you can raise an error if the input is a pulse that already contains a shape).

Reply:
>  This function is meant to be used only to go from rfsoc_pulses.Pulse to whatever pulses, right?

Yes exactly!
> eventually you can raise an error if the input is a pulse that already contains a shape

I mean, since I created the Pulse object just two lines before I don't think this is necessary. 
I added the test also because of that, to check everything was working :-)